text/0000-non-exhaustive.md

```diff
@@ -0,0 +1,548 @@
+ensure that an appropriate description is given for every variant. In fact, if
+they use a wildcard arm in addition to the existing variants, it should be
+identified as dead code, because it will never be run.
+
+Outside the crate that defines the enum, users should be required to add a
+wildcard arm to ensure forward-compatibility, like so:
+
+```rust
+use mycrate::Error;
+
+match error {
+    Message(ref s) => ...,
+    Other => ...,
+    _ => ...,
+}
+```
+
+And it should *not* be marked as dead code, even if the compiler does mark it as
+dead and remove it.
+
+Note that this can *potentially* cause breaking changes if a user adds
+`#[deny(dead_code)]` to a match statement *and* the upstream crate removes the
+`#[non_exhaustive]` lint. That said, modifying warn-only lints is generally
+assumed to not be a breaking change, even though users can make it a breaking
+change by manually denying lints.
+
+## Structs
+
+Like with enums, the attribute is essentially ignored in the crate that defines
+the struct, so that users can continue to construct values for the struct.
+However, this will prevent downstream users from constructing or exhaustively
+matching the struct, because fields may be added to the struct in the future.
+
+Additionally, adding `#[non_exhaustive]` to an enum variant will operate exactly
+the same as if the variant were a struct.
+
+Using our `Config` again:
+
+```rust
+#[non_exhaustive]
+pub struct Config {
+    pub window_width: u16,
+    pub window_height: u16,
+}
+```
+
+We can still construct our config within the defining crate like so:
+
+```rust
+let config = Config { window_width: 640, window_height: 480 };
+```
+
+And we can even exhaustively match on it, like so:
+
+```rust
+if let Ok(Config { window_width, window_height }) = load_config() {
+    // ...
+}
+```
+
+But users outside the crate won't be able to construct their own values, because
+otherwise, adding extra fields would be a breaking change.
+
+Users can still match on `Config`s non-exhaustively, as usual:
+
+```rust
+let &Config { window_width, window_height, .. } = config;
+```
+
+But without the `..`, this code will fail to compile.
+
+Although it should not be explicitly forbidden by the language to mark a struct
+with some private fields as non-exhaustive, it should emit a warning to tell the
+user that the attribute has no effect.
+
+## Tuple structs
+
+Non-exhaustive tuple structs will operate similarly to structs, however, will
+disallow matching directly. For example, take this example on stable today:
+
+```rust
+pub Config(pub u16, pub u16, ());
+```
+
+The below code does not work, because you can't match tuple structs with private
+fields:
+
+```rust
+let Config(width, height, ..) = config;
+```
+
+However, this code *does* work:
+
+```rust
+let Config { 0: width, 1: height, .. } = config;
+```
+
+So, if we label a struct non-exhaustive:
+
+```
+#[non_exhaustive]
+pub Config(pub u16, pub u16)
```

Comment:
This is the most interesting part of the RFC.

I suppose that using `Config` constructor outside of the crate will result in errors:
```rust
let Config(a, ..) = cfg; // error, Config is non-exhaustive, can't use
let Config(..) = cfg; // error, Config is non-exhaustive, can't use
let cfg = Config(a, b); // error, Config is non-exhaustive, can't use
let fn_ptr = Config; // error, Config is non-exhaustive, can't use
```
In this case this RFC is trying to resurrect hacks removed in https://github.com/rust-lang/rust/pull/38932 - the constructor is essentially unusable, but it still occupies a slot in the value namespace, can be reexported and cause minor disruptions.

What you actually want is to adjust the constructor's visibility when `#[non_exhaustive]` is specified.
```
vis(ctor) -> min(pub(crate), vis(ctor))
```